# *Data Preparation with Python* - Bagian 3
Pertanyaan, koreksi, maupun saran dapat disampaikan ke galuh.tunggadewi[at]gmail.com atau disampaikan dengan cara membuka [isu baru di GitHub](https://github.com/galuhsahid/data-preparation-with-python). Lisensi dapat dilihat [di sini](https://creativecommons.org/licenses/by-nc-sa/4.0/).

In [2]:
import pandas as pd

# Mengubah bentuk data

In [96]:
df_gdp = pd.read_csv("data/world_gdp.tsv", sep="\t")

Terkadang, kita mendapatkan data yang memiliki banyak kolom seperti data berikut:

In [97]:
df_gdp.head()

Country Name Country Code      Continent         Indicator Name  \
0        Aruba          ABW  North America  GDP growth (annual %)   
1  Afghanistan          AFG           Asia  GDP growth (annual %)   
2       Angola          AGO         Africa  GDP growth (annual %)   
3      Albania          ALB         Europe  GDP growth (annual %)   
4      Andorra          AND         Europe  GDP growth (annual %)   

      Indicator Code  1960  1961  1962  1963  1964  ...       2010      2011  \
0  NY.GDP.MKTP.KD.ZG   NaN   NaN   NaN   NaN   NaN  ...  -3.685030  3.446055   
1  NY.GDP.MKTP.KD.ZG   NaN   NaN   NaN   NaN   NaN  ...  14.362441  0.426355   
2  NY.GDP.MKTP.KD.ZG   NaN   NaN   NaN   NaN   NaN  ...   4.859220  3.471981   
3  NY.GDP.MKTP.KD.ZG   NaN   NaN   NaN   NaN   NaN  ...   3.706881  2.545405   
4  NY.GDP.MKTP.KD.ZG   NaN   NaN   NaN   NaN   NaN  ...  -5.358826 -4.646543   

        2012      2013      2014      2015      2016      2017      2018  \
0  -1.369863  4.198232  0.848228 -0.450586 -0.211225  1.330511       NaN   
1  12.752287  5.600745  2.724543  1.451315  2.260314  2.665292  1.030660   
2   8.542147  4.954590  4.822626  0.943576 -2.580097 -0.147207 -2.133493   
3   1.417526  1.001988  1.770000  2.230000  3.350000  3.836620  4.004413   
4  -1.615218  0.351645  2.277683  0.842204  1.889124  1.724022  1.629345   

   Unnamed: 64  
0          NaN  
1          NaN  
2          NaN  
3          NaN  
4          NaN  

[5 rows x 65 columns]

Pada contoh data di atas, untuk setiap tahunnya, terdapat kolom yang berisi GDP *growth* di tahun tersebut. Seringkali, kita ingin mengubah bentuk data tersebut, misalnya kita bisa memiliki kolom tahun sendiri dan kolom GDP *growth* yang berisi GDP *growth* di masing-masing negara dan tahun:


| Country Name | Country Code | ... | Year | GDP *growth* (%)

| Aruba | ABW | ... | 2011 | 3.446055

| Aruba | ABW | ... | 2012 | -1.369863

...

| Afghanistan | AFG | ... | 2011 | 3446055

## melt

📖  Dokumentasi `pandas.melt`: [tautan](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.melt.html)

- `id_vars`: *List* berisi nama-nama kolom yang kita tetapkan sebagai **variabel**. Kolom-kolom ini tidak akan berubah, tetap seperti semula.
- `value_vars` (opsional): *List* berisi nama-nama kolom yang ingin kita *unpivot*. Apabila tidak diisi, maka akan dianggap bahwa semua kolom yang tidak ditulis di `id_vars` akan di-*unpivot*.

Pada kode di bawah, kita tidak akan merubah kolom-kolom Country Name, Country Code, Continent, Indicator Name, dan Indicator Name. Kita akan meng-*unpivot* kolom-kolom lainnya, yang merupakan kolom-kolom tahun dari 1960 sampai seterusnya.

In [98]:
df_gdp_melt = pd.melt(df_gdp, id_vars=["Country Name", "Country Code", "Continent", "Indicator Name", "Indicator Code"])

In [99]:
df_gdp_melt.head()

Country Name Country Code      Continent         Indicator Name  \
0        Aruba          ABW  North America  GDP growth (annual %)   
1  Afghanistan          AFG           Asia  GDP growth (annual %)   
2       Angola          AGO         Africa  GDP growth (annual %)   
3      Albania          ALB         Europe  GDP growth (annual %)   
4      Andorra          AND         Europe  GDP growth (annual %)   

      Indicator Code variable  value  
0  NY.GDP.MKTP.KD.ZG     1960    NaN  
1  NY.GDP.MKTP.KD.ZG     1960    NaN  
2  NY.GDP.MKTP.KD.ZG     1960    NaN  
3  NY.GDP.MKTP.KD.ZG     1960    NaN  
4  NY.GDP.MKTP.KD.ZG     1960    NaN

Kita dapat menggunakan `sort_values()` untuk mengurutkan baris sesuai dengan negara terlebih dahulu, kemudian tahun, untuk mengecek apakah semua kolom tahun sudah di-*unpivot*.

In [100]:
df_gdp_melt = df_gdp_melt.sort_values(by=["Country Name", "variable"])

In [101]:
df_gdp_melt.head(200)

Country Name Country Code Continent         Indicator Name  \
1         Afghanistan          AFG      Asia  GDP growth (annual %)   
265       Afghanistan          AFG      Asia  GDP growth (annual %)   
529       Afghanistan          AFG      Asia  GDP growth (annual %)   
793       Afghanistan          AFG      Asia  GDP growth (annual %)   
1057      Afghanistan          AFG      Asia  GDP growth (annual %)   
1321      Afghanistan          AFG      Asia  GDP growth (annual %)   
1585      Afghanistan          AFG      Asia  GDP growth (annual %)   
1849      Afghanistan          AFG      Asia  GDP growth (annual %)   
2113      Afghanistan          AFG      Asia  GDP growth (annual %)   
2377      Afghanistan          AFG      Asia  GDP growth (annual %)   
2641      Afghanistan          AFG      Asia  GDP growth (annual %)   
2905      Afghanistan          AFG      Asia  GDP growth (annual %)   
3169      Afghanistan          AFG      Asia  GDP growth (annual %)   
3433      Afghanistan          AFG      Asia  GDP growth (annual %)   
3697      Afghanistan          AFG      Asia  GDP growth (annual %)   
3961      Afghanistan          AFG      Asia  GDP growth (annual %)   
4225      Afghanistan          AFG      Asia  GDP growth (annual %)   
4489      Afghanistan          AFG      Asia  GDP growth (annual %)   
4753      Afghanistan          AFG      Asia  GDP growth (annual %)   
5017      Afghanistan          AFG      Asia  GDP growth (annual %)   
5281      Afghanistan          AFG      Asia  GDP growth (annual %)   
5545      Afghanistan          AFG      Asia  GDP growth (annual %)   
5809      Afghanistan          AFG      Asia  GDP growth (annual %)   
6073      Afghanistan          AFG      Asia  GDP growth (annual %)   
6337      Afghanistan          AFG      Asia  GDP growth (annual %)   
6601      Afghanistan          AFG      Asia  GDP growth (annual %)   
6865      Afghanistan          AFG      Asia  GDP growth (annual %)   
7129      Afghanistan          AFG      Asia  GDP growth (annual %)   
7393      Afghanistan          AFG      Asia  GDP growth (annual %)   
7657      Afghanistan          AFG      Asia  GDP growth (annual %)   
...               ...          ...       ...                    ...   
13258         Algeria          DZA    Africa  GDP growth (annual %)   
13522         Algeria          DZA    Africa  GDP growth (annual %)   
13786         Algeria          DZA    Africa  GDP growth (annual %)   
14050         Algeria          DZA    Africa  GDP growth (annual %)   
14314         Algeria          DZA    Africa  GDP growth (annual %)   
14578         Algeria          DZA    Africa  GDP growth (annual %)   
14842         Algeria          DZA    Africa  GDP growth (annual %)   
15106         Algeria          DZA    Africa  GDP growth (annual %)   
15370         Algeria          DZA    Africa  GDP growth (annual %)   
15634         Algeria          DZA    Africa  GDP growth (annual %)   
9      American Samoa          ASM       NaN  GDP growth (annual %)   
273    American Samoa          ASM       NaN  GDP growth (annual %)   
537    American Samoa          ASM       NaN  GDP growth (annual %)   
801    American Samoa          ASM       NaN  GDP growth (annual %)   
1065   American Samoa          ASM       NaN  GDP growth (annual %)   
1329   American Samoa          ASM       NaN  GDP growth (annual %)   
1593   American Samoa          ASM       NaN  GDP growth (annual %)   
1857   American Samoa          ASM       NaN  GDP growth (annual %)   
2121   American Samoa          ASM       NaN  GDP growth (annual %)   
2385   American Samoa          ASM       NaN  GDP growth (annual %)   
2649   American Samoa          ASM       NaN  GDP growth (annual %)   
2913   American Samoa          ASM       NaN  GDP growth (annual %)   
3177   American Samoa          ASM       NaN  GDP growth (annual %)   
3441   American Samoa          ASM       NaN  GDP growth (annual %)   
3705   American Samoa          AS

Pada tabel di atas, kolom tahun kita bernama "variable" dan kolom GDP *growth* kita bernama "value". Nama-nama kolom ini dapat diubah dengan cara mendefinisikan `var_name` dan `value_name`.

In [102]:
df_gdp_melt = pd.melt(df_gdp, 
                      id_vars=["Country Name", "Country Code", "Continent", "Indicator Name", "Indicator Code"],
                      var_name='Year',
                      value_name='GDP growth (annual %)')

In [103]:
df_gdp_melt.head()

Country Name Country Code      Continent         Indicator Name  \
0        Aruba          ABW  North America  GDP growth (annual %)   
1  Afghanistan          AFG           Asia  GDP growth (annual %)   
2       Angola          AGO         Africa  GDP growth (annual %)   
3      Albania          ALB         Europe  GDP growth (annual %)   
4      Andorra          AND         Europe  GDP growth (annual %)   

      Indicator Code  Year  GDP growth (annual %)  
0  NY.GDP.MKTP.KD.ZG  1960                    NaN  
1  NY.GDP.MKTP.KD.ZG  1960                    NaN  
2  NY.GDP.MKTP.KD.ZG  1960                    NaN  
3  NY.GDP.MKTP.KD.ZG  1960                    NaN  
4  NY.GDP.MKTP.KD.ZG  1960                    NaN

### Latihan
Baca *file* berjudul "world_population.csv". Simpan di variabel bernama `df_pop`. Kemudian, buatlah tabel yang serupa dengan tabel `df_gdp_melt`. Simpan tabel di variabel `df_pop_melt`.

#### Solusi

In [104]:
df_pop = pd.read_csv("data/world_population.csv")

In [105]:
df_pop.head()

Country Name Country Code     Indicator Name Indicator Code       1960  \
0        Aruba          ABW  Population, total    SP.POP.TOTL    54211.0   
1  Afghanistan          AFG  Population, total    SP.POP.TOTL  8996973.0   
2       Angola          AGO  Population, total    SP.POP.TOTL  5454933.0   
3      Albania          ALB  Population, total    SP.POP.TOTL  1608800.0   
4      Andorra          AND  Population, total    SP.POP.TOTL    13411.0   

        1961       1962       1963       1964       1965  ...        2010  \
0    55438.0    56225.0    56695.0    57032.0    57360.0  ...    101669.0   
1  9169410.0  9351441.0  9543205.0  9744781.0  9956320.0  ...  29185507.0   
2  5531472.0  5608539.0  5679458.0  5735044.0  5770570.0  ...  23356246.0   
3  1659800.0  1711319.0  1762621.0  1814135.0  1864791.0  ...   2913021.0   
4    14375.0    15370.0    16412.0    17469.0    18549.0  ...     84449.0   

         2011        2012        2013        2014        2015        2016  \
0    102046.0    102560.0    103159.0    103774.0    104341.0    104872.0   
1  30117413.0  31161376.0  32269589.0  33370794.0  34413603.0  35383128.0   
2  24220661.0  25107931.0  26015780.0  26941779.0  27884381.0  28842484.0   
3   2905195.0   2900401.0   2895092.0   2889104.0   2880703.0   2876101.0   
4     83747.0     82427.0     80774.0     79213.0     78011.0     77297.0   

         2017        2018  Unnamed: 63  
0    105366.0    105845.0          NaN  
1  36296400.0  37172386.0          NaN  
2  29816748.0  30809762.0          NaN  
3   2873457.0   2866376.0          NaN  
4     77001.0     77006.0          NaN  

[5 rows x 64 columns]

In [106]:
df_pop_melt = pd.melt(df_pop, 
                      id_vars=["Country Name", "Country Code", "Indicator Name", "Indicator Code"],
                      var_name='Year',
                      value_name='Population')

In [107]:
df_pop_melt.head()

Country Name Country Code     Indicator Name Indicator Code  Year  \
0        Aruba          ABW  Population, total    SP.POP.TOTL  1960   
1  Afghanistan          AFG  Population, total    SP.POP.TOTL  1960   
2       Angola          AGO  Population, total    SP.POP.TOTL  1960   
3      Albania          ALB  Population, total    SP.POP.TOTL  1960   
4      Andorra          AND  Population, total    SP.POP.TOTL  1960   

   Population  
0     54211.0  
1   8996973.0  
2   5454933.0  
3   1608800.0  
4     13411.0

# Menggabungkan dua sumber data berbeda

## Sekilas mengenai operasi `join`

![alt text](https://placeisimportant.files.wordpress.com/2013/02/sql-joins-vis-rep-1.png)

[Sumber](https://placeisimportant.wordpress.com)

## Menggabungkan dua tabel dengan `pandas.DataFrame.merge`

📖  Dokumentasi `pandas.DataFrame.merge`: [tautan](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html)

Apabila Anda sudah cukup familiar dengan SQL, konsep `merge` di pandas mirip dengan konsep `join` di SQL.

- `on`: sebuah kolom atau sebuah *list* berisi nama-nama kolom yang akan digunakan sebagai acuan untuk menggabungkan tabel.
- `how="inner"` (*default*): menghasilkan baris-baris yang kuncinya ada pada kedua tabel. Serupa dengan *inner join* pada SQL.
- `how="left"`: menghasilkan baris-baris yang kuncinya ada pada tabel di sebelah kiri. Apabila kunci di tabel sebelah kiri tidak dapat ditemukan di tabel sebelah kanan, maka nilai untuk kolom-kolom dari tabel di sebelah kanan untuk kunci tersebut akan bernilai *null*. Serupa dengan *left outer join* pada SQL.
- `how="right"`: menghasilkan baris-baris yang kuncinya ada pada tabel di sebelah kanan. Apabila kunci di tabel sebelah kanan tidak dapat ditemukan di tabel sebelah kanan, maka nilai untuk kolom-kolom dari tabel di sebelah kiri untuk kunci tersebut akan bernilai *null*. Serupa dengan *right outer join* pada SQL.
- `how="outer"`: menghasilkan seluruh baris dengan kunci yang ada di kedua tabel. Serupa dengan *full outer join* di SQL.

Menggabungkan dua tabel secara *inner join* pada kolom "Country Name" dan "Year":

In [108]:
df_gdp_pop = df_gdp_melt.merge(df_pop_melt, how="inner", on=["Country Name", "Year"])

In [109]:
df_gdp_pop.head()

Country Name Country Code_x      Continent       Indicator Name_x  \
0        Aruba            ABW  North America  GDP growth (annual %)   
1  Afghanistan            AFG           Asia  GDP growth (annual %)   
2       Angola            AGO         Africa  GDP growth (annual %)   
3      Albania            ALB         Europe  GDP growth (annual %)   
4      Andorra            AND         Europe  GDP growth (annual %)   

    Indicator Code_x  Year  GDP growth (annual %) Country Code_y  \
0  NY.GDP.MKTP.KD.ZG  1960                    NaN            ABW   
1  NY.GDP.MKTP.KD.ZG  1960                    NaN            AFG   
2  NY.GDP.MKTP.KD.ZG  1960                    NaN            AGO   
3  NY.GDP.MKTP.KD.ZG  1960                    NaN            ALB   
4  NY.GDP.MKTP.KD.ZG  1960                    NaN            AND   

    Indicator Name_y Indicator Code_y  Population  
0  Population, total      SP.POP.TOTL     54211.0  
1  Population, total      SP.POP.TOTL   8996973.0  
2  Population, total      SP.POP.TOTL   5454933.0  
3  Population, total      SP.POP.TOTL   1608800.0  
4  Population, total      SP.POP.TOTL     13411.0

Kode di atas tidak akan mengembalikan negara-negara dan tahun yang tidak muncul di tabel populasi maupun tabel GDP.

In [110]:
df_gdp_pop = df_gdp_pop[["Country Name", "Continent", "Year", "GDP growth (annual %)", "Population"]]

Tabel `df_gdp_pop` merupakan gabungan kedua tabel, di mana baik GDP maupun populasi dari masing-masing negara dan tahun sudah terdapat pada satu tabel.

In [111]:
df_gdp_pop.head()

Country Name      Continent  Year  GDP growth (annual %)  Population
0        Aruba  North America  1960                    NaN     54211.0
1  Afghanistan           Asia  1960                    NaN   8996973.0
2       Angola         Africa  1960                    NaN   5454933.0
3      Albania         Europe  1960                    NaN   1608800.0
4      Andorra         Europe  1960                    NaN     13411.0

# Meringkas data

Terkadang kita ingin meringkas data, misalnya membuat tabel yang hanya berisi rata-rata GDP *growth* dari masing-masing negara dari tahun 1960. Kita dapat menggunakan `groupby()` untuk melakukan hal tersebut.

 ## `pandas.DataFrame.groupby`

📖  Dokumentasi `pandas.DataFrame.groupby`: [tautan](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html)

Pada contoh di bawah, kita akan menghitung rata-rata GDP *growth* di masing-masing negara:

In [113]:
df_gdp_pop.groupby(['Country Name'])["GDP growth (annual %)"].mean().head()

Country Name
Afghanistan       6.828081
Albania           2.969914
Algeria           3.765577
American Samoa   -1.354800
Andorra           2.621715
Name: GDP growth (annual %), dtype: float64

In [114]:
df_gdp_pop.groupby(['Country Name'])[["GDP growth (annual %)"]].mean()

GDP growth (annual %)
Country Name                                   
Afghanistan                            6.828081
Albania                                2.969914
Algeria                                3.765577
American Samoa                        -1.354800
Andorra                                2.621715
Angola                                 3.800455
Antigua and Barbuda                    3.847617
Arab World                             3.786073
Argentina                              2.495562
Armenia                                3.342892
Aruba                                  3.322736
Australia                              3.462451
Austria                                2.774991
Azerbaijan                             4.197467
Bahamas, The                           3.227790
Bahrain                                4.061887
Bangladesh                             4.271062
Barbados                               1.418358
Belarus                                2.670841
Belgium                                2.612755
Belize                                 5.048076
Benin                                  3.725667
Bermuda                                2.812817
Bhutan                                 7.394948
Bolivia                                3.664155
Bosnia and Herzegovina                 9.753138
Botswana                               8.190928
Brazil                                 4.012233
British Virgin Islands                      NaN
Brunei Darussalam                      1.656578
...                                         ...
Syrian Arab Republic                   5.804991
Tajikistan                             1.923856
Tanzania                               5.236051
Thailand                               5.990697
Timor-Leste                            8.069286
Togo                                   4.100170
Tonga                                  1.961722
Trinidad and Tobago                    3.334910
Tunisia                                4.504333
Turkey                                 4.752196
Turkmenistan                           5.599602
Turks and Caicos Islands               3.633933
Tuvalu                                 2.554945
Uganda                                 5.799064
Ukraine                               -1.080411
United Arab Emirates                   4.863038
United Kingdom                         2.414300
United States                          3.043988
Upper middle income                    4.660936
Uruguay                                2.331899
Uzbekistan                             4.423526
Vanuatu                                3.049852
Venezuela, RB                          2.805467
Vietnam                                6.453346
Virgin Islands (U.S.)                 -1.424620
West Bank and Gaza                     4.536008
World                                  3.487314
Yemen, Rep.                            1.959624
Zambia                                 3.331390
Zimbabwe                               3.123528

[262 rows x 1 columns]

Total GDP di masing-masing kontinen:

In [115]:
df_gdp_pop.groupby(['Continent'])[["GDP growth (annual %)"]].sum().head()

GDP growth (annual %)
Continent                           
Africa                   9343.615899
Asia                     8859.353980
Europe                   5501.990083
North America            4211.810820
Oceania                  1540.987390

### Latihan
1. Berapakah rata-rata GDP untuk setiap kontinen per tahun?

Misal:
- Asia | 1998 | 98492
- Asia | 1999 | 28374
- Eropa | 1998 | 290384


2. Tampilkan rata-rata populasi per tahun untuk negara-negara yang ada di benua Asia dan Eropa

#### Solusi 1

In [116]:
df_gdp_pop.groupby(['Continent', 'Year'])[["GDP growth (annual %)"]].mean()

GDP growth (annual %)
Continent     Year                       
Africa        1960                    NaN
              1961               2.314596
              1962               3.791991
              1963               4.951806
              1964               5.159206
              1965               4.959958
              1966               3.336221
              1967               3.678482
              1968               4.194378
              1969               5.231270
              1970               7.116509
              1971               5.462373
              1972               5.222336
              1973               2.894688
              1974               6.172224
              1975               3.774772
              1976               7.190731
              1977               4.571059
              1978               4.369473
              1979               3.563579
              1980               1.803749
              1981               2.874986
              1982               1.976954
              1983               0.803953
              1984               2.217047
              1985               4.690759
              1986               3.110291
              1987               2.911545
              1988               4.687020
              1989               3.200949
...                                   ...
South America 1989               0.058034
              1990               0.725018
              1991               4.812366
              1992               4.145724
              1993               3.806403
              1994               5.452640
              1995               3.700649
              1996               3.449540
              1997               5.763731
              1998               1.874072
              1999              -1.505568
              2000               1.526983
              2001               1.122907
              2002              -0.239217
              2003               2.552077
              2004               6.840514
              2005               5.080129
              2006               5.961884
              2007               6.244970
              2008               5.209109
              2009               0.497407
              2010               5.882656
              2011               5.623716
              2012               3.634390
              2013               4.495766
              2014               2.035457
              2015               1.441772
              2016               0.829382
              2017               2.432936
              2018               2.323712

[354 rows x 1 columns]

In [117]:
df_gdp_pop.groupby(['Continent', 'Year'], as_index=False)[["GDP growth (annual %)"]].mean()

Continent  Year  GDP growth (annual %)
0           Africa  1960                    NaN
1           Africa  1961               2.314596
2           Africa  1962               3.791991
3           Africa  1963               4.951806
4           Africa  1964               5.159206
5           Africa  1965               4.959958
6           Africa  1966               3.336221
7           Africa  1967               3.678482
8           Africa  1968               4.194378
9           Africa  1969               5.231270
10          Africa  1970               7.116509
11          Africa  1971               5.462373
12          Africa  1972               5.222336
13          Africa  1973               2.894688
14          Africa  1974               6.172224
15          Africa  1975               3.774772
16          Africa  1976               7.190731
17          Africa  1977               4.571059
18          Africa  1978               4.369473
19          Africa  1979               3.563579
20          Africa  1980               1.803749
21          Africa  1981               2.874986
22          Africa  1982               1.976954
23          Africa  1983               0.803953
24          Africa  1984               2.217047
25          Africa  1985               4.690759
26          Africa  1986               3.110291
27          Africa  1987               2.911545
28          Africa  1988               4.687020
29          Africa  1989               3.200949
..             ...   ...                    ...
324  South America  1989               0.058034
325  South America  1990               0.725018
326  South America  1991               4.812366
327  South America  1992               4.145724
328  South America  1993               3.806403
329  South America  1994               5.452640
330  South America  1995               3.700649
331  South America  1996               3.449540
332  South America  1997               5.763731
333  South America  1998               1.874072
334  South America  1999              -1.505568
335  South America  2000               1.526983
336  South America  2001               1.122907
337  South America  2002              -0.239217
338  South America  2003               2.552077
339  South America  2004               6.840514
340  South America  2005               5.080129
341  South America  2006               5.961884
342  South America  2007               6.244970
343  South America  2008               5.209109
344  South America  2009               0.497407
345  South America  2010               5.882656
346  South America  2011               5.623716
347  South America  2012               3.634390
348  South America  2013               4.495766
349  South America  2014               2.035457
350  South America  2015               1.441772
351  South America  2016               0.829382
352  South America  2017               2.432936
353  South America  2018               2.323712

[354 rows x 3 columns]

#### Solusi 2

In [118]:
df_gdp_pop[df_gdp_pop["Continent"].isin(["Europe", "Asia"])].groupby(['Country Name', 'Year'])[["Population"]].mean()

Population
Country Name Year            
Afghanistan  1960   8996973.0
             1961   9169410.0
             1962   9351441.0
             1963   9543205.0
             1964   9744781.0
             1965   9956320.0
             1966  10174836.0
             1967  10399926.0
             1968  10637063.0
             1969  10893776.0
             1970  11173642.0
             1971  11475445.0
             1972  11791215.0
             1973  12108963.0
             1974  12412950.0
             1975  12689160.0
             1976  12943093.0
             1977  13171306.0
             1978  13341198.0
             1979  13411056.0
             1980  13356511.0
             1981  13171673.0
             1982  12882528.0
             1983  12537730.0
             1984  12204292.0
             1985  11938208.0
             1986  11736179.0
             1987  11604534.0
             1988  11618005.0
             1989  11868877.0
...                       ...
Yemen, Rep.  1989  11189177.0
             1990  11709993.0
             1991  12302124.0
             1992  12954155.0
             1993  13634076.0
             1994  14297613.0
             1995  14913315.0
             1996  15469274.0
             1997  15975668.0
             1998  16450310.0
             1999  16921149.0
             2000  17409072.0
             2001  17918373.0
             2002  18443691.0
             2003  18985000.0
             2004  19540098.0
             2005  20107409.0
             2006  20687646.0
             2007  21282515.0
             2008  21892146.0
             2009  22516460.0
             2010  23154855.0
             2011  23807588.0
             2012  24473178.0
             2013  25147109.0
             2014  25823485.0
             2015  26497889.0
             2016  27168210.0
             2017  27834821.0
             2018  28498687.0

[4838 rows x 1 columns]

## Agregasi

Kita juga dapat menggunakan `agg()` apabila kita ingin menampilkan lebih dari satu jenis perhitungan di tabel yang sama (misalnya, total jumlah populasi dan rata-rata GDP per tahun untuk masing-masing negara).

In [120]:
df_gdp_pop.groupby(['Continent', 'Year']).agg(
{
    'Population': sum,
    'GDP growth (annual %)': "mean"
})

Population  GDP growth (annual %)
Continent     Year                                    
Africa        1960  265176766.0                    NaN
              1961  271580213.0               2.314596
              1962  278219665.0               3.791991
              1963  285089394.0               4.951806
              1964  292179388.0               5.159206
              1965  299485430.0               4.959958
              1966  307011761.0               3.336221
              1967  314772260.0               3.678482
              1968  322785168.0               4.194378
              1969  331073710.0               5.231270
              1970  339659599.0               7.116509
              1971  348551190.0               5.462373
              1972  357761150.0               5.222336
              1973  367312418.0               2.894688
              1974  377233452.0               6.172224
              1975  387546522.0               3.774772
              1976  398257947.0               7.190731
              1977  409371909.0               4.571059
              1978  420896020.0               4.369473
              1979  432849950.0               3.563579
              1980  445233967.0               1.803749
              1981  458040638.0               2.874986
              1982  471256615.0               1.976954
              1983  484869705.0               0.803953
              1984  498869787.0               2.217047
              1985  513232093.0               4.690759
              1986  527967413.0               3.110291
              1987  543053622.0               2.911545
              1988  558389129.0               4.687020
              1989  573853856.0               3.200949
...                         ...                    ...
South America 1989  289829995.0               0.058034
              1990  295280439.0               0.725018
              1991  300706713.0               4.812366
              1992  306101381.0               4.145724
              1993  311470189.0               3.806403
              1994  316822704.0               5.452640
              1995  322162422.0               3.700649
              1996  327490755.0               3.449540
              1997  332794491.0               5.763731
              1998  338045088.0               1.874072
              1999  343204899.0              -1.505568
              2000  348245826.0               1.526983
              2001  353160147.0               1.122907
              2002  357951719.0              -0.239217
              2003  362619722.0               2.552077
              2004  367167054.0               6.840514
              2005  371598749.0               5.080129
              2006  375909963.0               5.961884
              2007  380105938.0               6.244970
              2008  384214554.0               5.209109
              2009  388272882.0               0.497407
              2010  392200507.0               5.882656
              2011  396272340.0               5.623716
              2012  400315907.0               3.634390
              2013  404321311.0               4.495766
              2014  408271864.0               2.035457
              2015  412155398.0               1.441772
              2016  415967283.0               0.829382
              2017  419712493.0               2.432936
              2018  423398995.0               2.323712

[354 rows x 2 columns]

### Latihan
Untuk masing-masing pasang benua dan tahun, tampilkan data:
- Min, max, dan mean populasi
- Rata-rata GDP *growth*
- Jumlah negara di benua tersebut

#### Solusi

In [123]:
df_gdp_pop.groupby(['Continent', 'Year']).agg(
{
    'Population': ["min", "max", "mean"],
    'GDP growth (annual %)': "mean",
    'Country Name': "count"
})

Population                             \
                          min          max          mean   
Continent     Year                                         
Africa        1960    41700.0   45138458.0  5.524516e+06   
              1961    42889.0   46063563.0  5.657921e+06   
              1962    44042.0   47029822.0  5.796243e+06   
              1963    45176.0   48032934.0  5.939362e+06   
              1964    46322.0   49066760.0  6.087071e+06   
              1965    47500.0   50127921.0  6.239280e+06   
              1966    48699.0   51217973.0  6.396078e+06   
              1967    49911.0   52342233.0  6.557755e+06   
              1968    51134.0   53506196.0  6.724691e+06   
              1969    52365.0   54717039.0  6.897369e+06   
              1970    53600.0   55982144.0  7.076242e+06   
              1971    54695.0   57296983.0  7.261483e+06   
              1972    56029.0   58665808.0  7.453357e+06   
              1973    56892.0   60114625.0  7.652342e+06   
              1974    57937.0   61677177.0  7.859030e+06   
              1975    59292.0   63374298.0  8.073886e+06   
              1976    60504.0   65221378.0  8.297041e+06   
              1977    61786.0   67203128.0  8.528581e+06   
              1978    62150.0   69271917.0  8.768667e+06   
              1979    62686.0   71361131.0  9.017707e+06   
              1980    63261.0   73423633.0  9.275708e+06   
              1981    64035.0   75440502.0  9.542513e+06   
              1982    64413.0   77427546.0  9.817846e+06   
              1983    64335.0   79414840.0  1.010145e+07   
              1984    64717.0   81448755.0  1.039312e+07   
              1985    65244.0   83562785.0  1.069234e+07   
              1986    65652.0   85766399.0  1.099932e+07   
              1987    68499.0   88048032.0  1.131362e+07   
              1988    68755.0   90395271.0  1.163311e+07   
              1989    69167.0   92788027.0  1.195529e+07   
...                       ...          ...           ...   
South America 1989   397085.0  146328304.0  2.415250e+07   
              1990   405168.0  149003223.0  2.460670e+07   
              1991   413005.0  151648011.0  2.505889e+07   
              1992   420655.0  154259380.0  2.550845e+07   
              1993   428030.0  156849078.0  2.595585e+07   
              1994   435105.0  159432716.0  2.640189e+07   
              1995   441851.0  162019896.0  2.684687e+07   
              1996   448213.0  164614688.0  2.729090e+07   
              1997   454161.0  167209040.0  2.773287e+07   
              1998   459831.0  169785250.0  2.817042e+07   
              1999   465385.0  172318675.0  2.860041e+07   
              2000   470949.0  174790340.0  2.902049e+07   
              2001   476579.0  177196054.0  2.943001e+07   
              2002   482235.0  179537520.0  2.982931e+07   
              2003   487942.0  181809246.0  3.021831e+07   
              2004   493679.0  184006481.0  3.059725e+07   
              2005   499464.0  186127103.0  3.096656e+07   
              2006   505295.0  188167356.0  3.132583e+07   
              2007   511181.0  190130443.0  3.167549e+07   
              2008   517123.0  192030362.0  3.201788e+07   
              2009   523111.0  193886508.0  3.235607e+07   
              2010   529131.0  195713635.0  3.268338e+07   
              2011   535179.0  197514534.0  3.302270e+07   
              2012   541245.0  199287296.0  3.335966e+07   
              2013   547291.0  201035903.0  3.369344e+07   
              2014   553273.0  202763735.0  3.402266e+07   
              2015   559143.0  204471769.0  3.434628e+07   
              2016   564888.0  206163058.0  3.466394e+07   
              2017   570496.0  207833831.0  3.497604e+07   
              2018   575991.0  209469333.0  3.528325e+07   

                   GDP growth (annual %) Country Name  
                                    mean        count  
Continent     Year                                     
Africa    

## Mengaplikasikan fungsi

In [124]:
df_aps = pd.read_csv("data/angka_partisipasi_sekolah.csv")

In [125]:
df_aps.head()

_id  idx                     vervar                              var  \
0    1    0  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
1    2    1  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
2    3    2  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
3    4    3  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
4    5    4  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   

          turvar  tahun  nilai  
0  Sekolah Dasar   2010  96.34  
1  Sekolah Dasar   2011  94.17  
2  Sekolah Dasar   2012  95.45  
3  Sekolah Dasar   2013    NaN  
4  Sekolah Dasar   2014  98.22

Kita dapat mengaplikasikan fungsi yang telah kita buat sendiri untuk menghasilkan nilai/kolom baru pada DataFrame kita. Sebagai contoh, kita ingin menghilangkan kata "Kabupaten" di kolom `vervar` apabila nilai di kolom tersebut diawali dengan kata "Kabupaten". Pertama-tama, buatlah fungsinya terlebih dahulu:

In [126]:
def get_kabupaten_name(value):
  if value.startswith("Kabupaten "): # apabila input diawali dengan kata "Kabupaten "
    return " ".join(value.split(" ")[1:]) # buat string menjadi list berisi kata-kata di string tersebut,
                                          # ambil elemen ke-1 sampai elemen terakhir, dan gabungkan lagi
                                          # list tersebut menjadi sebuah string.
  return value

Cek apakah fungsi sudah berjalan dengan baik atau belum:

In [127]:
get_kabupaten_name("Kabupaten Polewali Mandar")

'Polewali Mandar'

Setelah dipastikan bahwa fungsi sudah berjalan dengan baik, gunakan `apply()` untuk mengaplikasikan fungsi tersebut secara otomatis ke kolom `vervar`.

In [128]:
df_aps["vervar"] = df_aps["vervar"].apply(get_kabupaten_name)

In [129]:
df_aps.head()

_id  idx           vervar                              var         turvar  \
0    1    0  Polewali Mandar  Angka Partisipasi Sekolah (APS)  Sekolah Dasar   
1    2    1  Polewali Mandar  Angka Partisipasi Sekolah (APS)  Sekolah Dasar   
2    3    2  Polewali Mandar  Angka Partisipasi Sekolah (APS)  Sekolah Dasar   
3    4    3  Polewali Mandar  Angka Partisipasi Sekolah (APS)  Sekolah Dasar   
4    5    4  Polewali Mandar  Angka Partisipasi Sekolah (APS)  Sekolah Dasar   

   tahun  nilai  
0   2010  96.34  
1   2011  94.17  
2   2012  95.45  
3   2013    NaN  
4   2014  98.22

Apabila kita ingin menempatkan hasil dari fungsi tersebut di kolom yang baru, kita bisa langsung mendefinisikan kolom baru tersebut:

In [130]:
df_aps = pd.read_csv("data/angka_partisipasi_sekolah.csv")

In [131]:
df_aps.head()

_id  idx                     vervar                              var  \
0    1    0  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
1    2    1  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
2    3    2  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
3    4    3  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
4    5    4  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   

          turvar  tahun  nilai  
0  Sekolah Dasar   2010  96.34  
1  Sekolah Dasar   2011  94.17  
2  Sekolah Dasar   2012  95.45  
3  Sekolah Dasar   2013    NaN  
4  Sekolah Dasar   2014  98.22

In [132]:
df_aps["vervar_new"] = df_aps["vervar"].apply(get_kabupaten_name)

In [133]:
df_aps.head()

_id  idx                     vervar                              var  \
0    1    0  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
1    2    1  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
2    3    2  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
3    4    3  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
4    5    4  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   

          turvar  tahun  nilai       vervar_new  
0  Sekolah Dasar   2010  96.34  Polewali Mandar  
1  Sekolah Dasar   2011  94.17  Polewali Mandar  
2  Sekolah Dasar   2012  95.45  Polewali Mandar  
3  Sekolah Dasar   2013    NaN  Polewali Mandar  
4  Sekolah Dasar   2014  98.22  Polewali Mandar

Bagaimana apabila kita membutuhkan lebih dari satu kolom untuk menghasilkan nilai baru? Misalnya, kita ingin menghitung luas dari masing-masing segi empat yang ada di tabel di bawah ini:

In [134]:
rectangles = [
    { 'height': 30, 'width': 4 },
    { 'height': 52, 'width': 6 },
    { 'height': 14, 'width': 3 }
]

rectangles_df = pd.DataFrame(rectangles)
rectangles_df

height  width
0      30      4
1      52      6
2      14      3

Buat fungsi untuk menghitung luas, dengan input berupa baris dari tabel tersebut:

In [135]:
def calculate_area(row):
    # sesuaikan dengan nama kolom, yaitu height dan width
    return row['height'] * row['width']

In [136]:
rectangles_df['area'] = rectangles_df.apply(calculate_area, axis=1)
rectangles_df

height  width  area
0      30      4   120
1      52      6   312
2      14      3    42

# *Checkpoint*
- Mengubah bentuk data
- Menggabungkan dua sumber data berbeda
- Meringkas data (groupby, agregasi)
- Mengaplikasikan fungsi yang dibuat sendiri ke tabel

![alt text](https://media2.giphy.com/media/d31w24psGYeekCZy/giphy.gif?cid=790b76115d4488ac643443597371b625&rid=giphy.gif)